project-squirrel-dragon/PresentationLayer/Modules/WatchStatsModule/WatchStatsViewModel.swift

```diff
@@ -0,0 +1,151 @@
+    func didFailWithError(message: String)
+}
+
+class WatchStatsViewModel {
+    typealias PokemonTCGCards = CardData
+    typealias FirebaseUserCards = UserCardData
+    typealias FirebaseUserSets = UserSetData
+
+    // MARK: - Properties
+    private weak var delegate: WatchStatsViewModelDelegate?
+    private var pokemonCardsRepository: TCGCardsRepositoryProtocol
+    private var userCardsRepository: RepositoryProtocol
+    private var userSetRepository: RepositoryProtocol
+
+    // MARK: - Other Properties
+    // TODO: - save rarity progress on firebase
+    private(set) var pokemonSetCards: [String:[String: PokemonTCGCards]] = [:]
+    private(set) var userPokemonCards: [String: FirebaseUserCards] = [:]
+    private(set) var setCollectableCards: [String: CardsOrganiser] = [:]
+
+    // MARK: - Initialization
+    init(delegate: WatchStatsViewModelDelegate,
+         pokemonCardsRepository: TCGCardsRepositoryProtocol,
+         userCardsRepository: RepositoryProtocol,
+         userSetRepository: RepositoryProtocol) {
+        self.delegate = delegate
+        self.pokemonCardsRepository = pokemonCardsRepository
+        self.userCardsRepository = userCardsRepository
+        self.userSetRepository = userSetRepository
+    }
+
+    // MARK: - Data Fetching Methods
+    func fetchViewData() {
+        // 1. fetch user sets
+        userSetRepository.fetch { [weak self] result in
+            switch result {
+            case .success(let userData):
+                guard let userSetsData = userData as? FirebaseData<FirebaseUserSets> else {
+                    self?.delegate?.didFailWithError(message: "Failed to cast data to UserSetData")
+                    return
+                }
+                userSetsData.data.forEach { self?.fetchSetCards(forSetId: $0.id) }
+
+                return
+            case .failure(let error):
+                self?.delegate?.didFailWithError(message: error.localizedDescription)
+                return
+            }
+        }
+    }
+
+    fileprivate func fetchSetCards(forSetId setId: String) {
+        // 2. fetch cards for user sets
+        pokemonCardsRepository.fetch(itemWithID:  "set.id:\(setId)") { [weak self] result in
+            switch result {
+            case .success(let data):
+                guard let pokemonCardsData = data as? PokemonCardsData else {
+                    self?.delegate?.didFailWithError(message: "Failed to cast data to PokemonCardsData")
+                    return
+                }
+
+                let cards = pokemonCardsData.data
+                self?.pokemonSetCards[setId] = Dictionary(uniqueKeysWithValues: cards.map { ($0.id, $0) })
+
+                self?.fetchUserCards()
+            case .failure(let error):
+                self?.delegate?.didFailWithError(message: error.localizedDescription)
+                return
+            }
+
+        }
+    }
+
+    fileprivate func fetchUserCards() {
+        // 3. fetch user cards for user sets
```

Comment:
Feel like these comments are not really necessary the function name is descriptive enough 😄